core/src/ecs/system/dispatcher.cpp

```diff
@@ -220,61 +250,126 @@ bool Dispatcher::dfsVisit(DFSNode& node, std::vector<DFSNode>& nodes)
-                    // Check if the condition returned true
-                    if (!mRetConditions.test(i))
+                    mRunConditions.set(i);
+                    if (mConditions[i].run(cmds))
                     {
-                        canRun = false;
-                        break;
+                        mRetConditions.set(i);
                     }
                 }
-
-                i += 1;
-                conditionsMask >>= 1;
+                // Check if the condition returned true
+                if (!mRetConditions.test(i))
+                {
+                    canRun = false;
+                    break;
+                }
             }
+
+            i += 1;
+            conditionsMask >>= 1;
         }
+    }
+
+    if (canRun)
+    {
+        mSystem->system.run(cmds);
+    }
+
+    // TODO: Check synchronization concerns when this gets multithreaded
+    cmds.commit();
+}
 
+void Dispatcher::GroupStep::call(CommandBuffer& cmds, std::vector<ecs::System<bool>>& mConditions,
+                                 std::bitset<CUBOS_CORE_DISPATCHER_MAX_CONDITIONS> mRunConditions,
+                                 std::bitset<CUBOS_CORE_DISPATCHER_MAX_CONDITIONS> mRetConditions)
+{
+    bool canRun = true;
+    while (canRun)
+    {
+        auto conditionsMask = conditions;
+        std::size_t i = 0;
+        while (conditionsMask.any())
+        {
+            if (conditionsMask.test(0))
+            {
+                // We have a condition, check if it has run already
+                if (!mRunConditions.test(i))
+                {
+                    mRunConditions.set(i);
+                }
+                if ((mConditions[i]).run(cmds))
+                {
+                    mRetConditions.set(i);
+                }
+                else
+                {
+                    mRetConditions.reset(i);
+                }
+                // Check if the condition returned true
+                if (!mRetConditions.test(i))
+                {
+                    canRun = false;
+                    break;
+                }
+            }
+
+            i += 1;
+            conditionsMask >>= 1;
+        }
         if (canRun)
         {
-            system->system.run(cmds);
+            for (auto& step : mSteps)
+            {
+                if (auto groupStepPtr = std::dynamic_pointer_cast<GroupStep>(step))
+                {
+                    // CUBOS_WARN("{}", groupStepPtr->grouptag);
+                    groupStepPtr->call(cmds, mConditions, mRunConditions, mRetConditions);
+                }
+                else if (auto systemStepPtr = std::dynamic_pointer_cast<SystemStep>(step))
+                {
+                    // CUBOS_WARN("{}", SystemStepPtr->getSystem()->grouptag);
+                    systemStepPtr->call(cmds, mConditions, mRunConditions, mRetConditions);
+                }
+            }
+        }
+        if (grouptag == "main")
+        {
+            canRun = false;
         }
```

Comment:
What is this?

Reply:
main is the only group that doesn't have a condition attached to it, if this wasn't there the loop  would repeat forever.